Support both polling and streaming logs

eddycharly opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

This PR adds support for streaming logs instead of polling.

The proxy now supports streaming responses, this allows us to stream logs instead of polling.
Polling is still the default, the user has to enable streaming to use it.

The `Logs` component has been reworked to adapt to both streaming and polling use cases.
A flag was introduced to allow the frontend switching to streaming when enabled.
The installer and overlays were updated.
Doc was updated with the new flag.

I reverted the changes i did with the log lines limit.

/cc @a-roberts @AlanGreene 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

eddycharly commented:
/hold
depends on #1576

eddycharly commented:
tests should be green ... 🤞

eddycharly commented:
I didn't think about the 30s kill, good point.

eddycharly commented:
I'm not sure the 30s thing is an issue, if i'm not mistaken, it's the load balancer that kills the request, not the ingress controller, and i feel like it should not kill this kind of request because it's only sending data (writing in the response stream).

At least that's what i observe on my AWS Application load balancer (idle timeout is 60s but the request keeps working)

AlanGreene commented:
We've seen this on dogfooding + robocat with the websocket connection being killed every 30s:
https://github.com/tektoncd/dashboard/issues/1021

I referenced the GKE Ingress settings for increasing the timeout in that issue and some threads in #plumbing, this is a common problem.

eddycharly commented:
I am thinking about adding a flag to make this optional.
Like `--stream-logs` in the backend to drive the frontend behavior, leaving it disabled by default, what do you think ?

eddycharly commented:
@AlanGreene I changed the approach, reverted the changes related to log lines limit and added a flag to enable streaming so that both polling and streaming are supported (polling being the default).

It should solve the issue with GKE ingress, the third party using the `Log` component should be able to continue using it without troubles, and search should be the same as before.

eddycharly commented:
/hold cancel

#1576 was merged

eddycharly commented:
Just rebased but the `needs-rebase` flag doesn't go away

AlanGreene commented on package.json:
```diff
@@ -52,7 +52,8 @@
     "redux-thunk": "^2.3.0",
-    "regenerator-runtime": "^0.13.5"
+    "regenerator-runtime": "^0.13.5",
+    "web-streams-polyfill": "^2.1.1"
```
this should be a dev dependency
`npm remove web-streams-polyfill && npm install -D web-streams-polyfill`

AlanGreene commented on packages/components/src/components/Log/Log.js:
```diff
@@ -92,27 +92,50 @@ export class LogContainer extends Component {
+          this.reader
+            .read()
+            .then(result => this.readChunks(result, decoder))
+            .catch(error => {
+              throw error;
+            });
```
this error isn't being handled
if you want it to be caught by the `catch` block below you'll need to do something like `await this.reader.read()...`

eddycharly replied:
Good catch !

eddycharly replied:
Fixed

AlanGreene commented on packages/components/src/components/Log/Log.js:
```diff
@@ -92,27 +92,50 @@ export class LogContainer extends Component {
+          }
+        }
+      } catch (error) {
+        console.log(error); // eslint-disable-line no-console
```
`console.error` would be appropriate here if we want to log it

AlanGreene commented on packages/components/src/components/Log/Log.js:
```diff
@@ -122,6 +145,9 @@ export class LogContainer extends Component {
           ]
         });
+        if (stepStatus && !stepStatus.terminated) {
+          this.timer = setTimeout(() => this.loadLog(), pollingInterval);
```
no need for the additional anonymous function here

```suggestion
          this.timer = setTimeout(this.loadLog, pollingInterval);
```

AlanGreene commented on src/api/index.js:
```diff
@@ -305,10 +305,16 @@ export function getCondition({ name, namespace }) {
+export function getPodLogURL({ container, name, namespace, follow }) {
   let queryParams;
-  if (container) {
-    queryParams = { container };
+  if (container || follow) {
+    queryParams = {};
+    if (container) {
+      queryParams.container = container;
+    }
+    if (follow) {
+      queryParams.follow = follow;
+    }
   }
```
This can be condensed as:

```js
const queryParams = {
  ...(container && { container }),
  ...(follow && { follow })
};
```

Empty object for queryParams should be treated the same as null / undefined so it's safe to do this regardless of the values of `container` or `follow`, and it'll only include the keys with a truthy value.

AlanGreene commented on src/api/comms.js:
```diff
@@ -91,14 +94,18 @@ export async function request(uri, options = defaultOptions) {
-    method: 'GET',
-    headers: getHeaders(headers)
-  });
+export function get(uri, headers, stream = false) {
```
Could we have an `options` params here so we can name this flag where it's used?

```js
export function get(uri, headers, options = {}) {
```

```diff
const uri = getPodLogURL({ container, name, namespace, follow: 1 });
-return get(uri, { Accept: 'text/plain' }, true);
+return get(uri, { Accept: 'text/plain' }, { stream: true });
```

AlanGreene commented on src/containers/TaskRun/TaskRun.js:
```diff
@@ -38,14 +38,16 @@ import {
   getTaskRunsErrorMessage,
-  isWebSocketConnected
+  isWebSocketConnected,
+  streamLogs
```
can we rename this to make it clear it's a boolean? `isLogStreamingEnabled` or something like that

eddycharly replied:
Fixed, thanks

AlanGreene commented on src/api/index.js:
```diff
@@ -323,6 +329,11 @@ export function getPodLog({ container, name, namespace }) {
   return get(uri, { Accept: 'text/plain' });
 }
 
+export function followPodLog({ container, name, namespace }) {
+  const uri = getPodLogURL({ container, name, namespace, follow: 1 });
+  return get(uri, { Accept: 'text/plain' }, true);
+}
```
could we combine this with `getPodLog` above?

```js
export function getPodLog({ container, name, namespace, stream }) {
  const uri = getPodLogURL({ container, name, namespace, follow: stream });
  return get(uri, { Accept: 'text/plain' }, { stream });
}
```

eddycharly commented:
@AlanGreene Please double check, but i think i fixed all your comments.

eddycharly commented:
/test all

eddycharly commented:
There's something going wrong with the tests :(

eddycharly commented:
/test all

eddycharly commented:
It's green ! 😅

eddycharly commented:
Gonna try, maybe i made a mistake when rebasing.

eddycharly commented:
Fixed, we exited too early (without setting the state to an empty string) when done came in the first chunk.

eddycharly commented:
/test pull-tekton-dashboard-build-tests

eddycharly commented:
/test tekton-dashboard-unit-tests

eddycharly commented:
/test all

eddycharly commented:
/test tekton-dashboard-unit-tests

eddycharly commented:
/test tekton-dashboard-unit-tests

eddycharly commented:
/retest

eddycharly commented:
/retest

eddycharly commented:
/test pull-tekton-dashboard-integration-0-13-2-tests

eddycharly commented:
/meow

AlanGreene commented:
Not seeing the hanging skeleton anymore but it's not displaying the 'No log available' message with `stream-logs` enabled.

eddycharly commented:
Ah, 'No logs available' is the expected behavior ?
I thought it was when we couldn't fetch logs only.

AlanGreene commented:
'Unable to fetch log' for the error case

eddycharly commented:
Ok, so we have to set state to loading = false and logs = undefined. Doing it now, thanks.

eddycharly commented:
Should be fixed now... hopefully 🤞

AlanGreene commented:
@eddycharly there's a merge conflict here.

I'm also seeing an issue with `stream-logs` enabled:
![image](https://user-images.githubusercontent.com/2829095/87288496-56c3f000-c4f3-11ea-8fff-eb0a7967a831.png)

I have a cancelled PipelineRun, and when I view the log tab for any of the cancelled steps it continually retries the `follow=true` request.


As far as I can tell all of the other cases seem to be working as expected.

eddycharly commented:
@AlanGreene just rebased.

Let's fix the canceling issue in another PR if it's ok. I will give a look to the issue once it's merged, it should not be hard, probably just a matter of checking the status.
